File: src/bin/c16_exercise.rs
use std::collections::HashMap;

pub fn normalise(input: &str) -> String {
    input.trim_matches(|c: char | !c.is_alphanumeric()).to_lowercase()
}

pub fn word_count(input: &str) -> HashMap<String, usize> {
    // TODO: Count word frequencies in `input`, where each word is:
    //   1) split by whitespace (.split_whitespace())
    //   2) normalized: trim non-alphanumeric, lowercase
    //   3) skipped if it becomes empty
    //   4) counted with the entry API
    let mut counts: HashMap<String, usize> = HashMap::new();
    for string in input.split_whitespace() {
        let procsesed_string = normalise(string);
        *counts.entry(procsesed_string).or_insert(0) += 1
    }

    counts
}

fn main() {
    let input = "Hello, hello world!";
    let counts = word_count(input);
    println!("{counts:?}");
}

File: python_exercises/c16_exercise.py
def normalise(word: str) -> str:
    stripped = word.strip()
    while stripped and not stripped[0].isalnum():
        stripped = stripped[1:]
    while stripped and not stripped[-1].isalnum():
        stripped = stripped[:-1]
    return stripped.lower()


def word_count(text: str) -> dict[str, int]:
    counts: dict[str, int] = {}
    for word in text.split():
        w = normalise(word)
        if w:
            counts[w] = counts.get(w, 0) + 1
    return counts
